# Functions advanced topics  函数高级话题

Lambda function 匿名函数

filter map reduce 过滤，映射，归约 

Iterator 迭代器

Generator 生成器



## Lambda function 匿名函数


匿名函数是没有名字的函数。

语法： ***lambda arguments: expression***


In [14]:
f = lambda x: x * x
f(4)

16

In [15]:
def f(x):
  return x * x
f(4)

16

以上两个函数的功能是一样。 

lambda 可以快速定义简单的单行函数。

In [16]:
def power(n):
  return lambda a: pow(a, n)

tenth_pow = power(10)

tenth_pow(2)


1024

lambda 可以用于返回值

上例中定义了一个偏函数（Partial function）

## filter map reduce 过滤，映射，归约

函数的参数可以是另一个函数，这种函数就称之为高阶函数

lambda 经常可以和filter() 等高阶函数结合使用，比如做序列解析（List comprehension）

In [17]:
old_list = [27, 38, 53, 41, 65, 368, 89, 572]

new_list = list(filter(lambda x: (x % 2 == 0) , old_list))

print(new_list)

[38, 368, 572]


filter() 接收一个函数和一个序列

上例中定义了一个判断偶数的lambda 

In [18]:
old_list = [27, 38, 53, 41, 65, 368, 89, 572]

new_list = list(map(lambda x: x % 2, old_list))
print(new_list)

another_list = list(map(str, old_list))
print(another_list)

additions = list(map("+".join, another_list))
print(additions)

[1, 0, 1, 1, 1, 0, 1, 0]
['27', '38', '53', '41', '65', '368', '89', '572']
['2+7', '3+8', '5+3', '4+1', '6+5', '3+6+8', '8+9', '5+7+2']


map() 也接收一个函数和一个序列

如果给一年级小学生出加法题。。。

In [19]:
old_list = [27, 38, 53, 41, 65, 368, 89, 572]

new_list = list(filter(lambda x: x < 100, old_list))
print(new_list)

from functools import reduce
sum1 = reduce(lambda a, b: a+b, new_list)
print(sum1)

import operator 
sum2 = reduce(operator.add, new_list)
print(sum2)

[27, 38, 53, 41, 65, 89]
313
313


如果给三年级小学生出加法题。。。

reduce() 也接收一个函数和一个序列

reduce() 将函数应用于所有序列元素 

上例中求和 (((((27 + 38) + 53) + 41) + 65) + 89)

注意要先从 functools 导入 reduce() 

如果要用add() 可以从 operator 导入

## Iterator 迭代器

可以不断返回下一个值的对象称为迭代器：Iterator

可以直接作用于for 循环的对象统称为可迭代对象：Iterable

In [20]:
it = iter("Je t'aime")
print(type(it))

while True:
    try:
        x = print(next(it))
    except StopIteration:
        break

<class 'str_iterator'>
J
e
 
t
'
a
i
m
e


iter() 把Iterable 变成 Iterator

对Iterator 操作，取下一个值用next()

如果迭代超出范围，就抛出错误：StopIteration

## Generator 生成器

生成器是一个函数，它返回一个Iterator 对象，可以对其进行迭代。

定义一个生成器，只要用yield 返回需要的值就行了。

生成器本质上还是一个迭代器


In [21]:
def fib(n):
    i, a, b = 0, 0, 1
    while i < n:
        print(b)
        a, b = b, a + b
        i += 1
    return 'The last one is F' + str(n)

fib(9)

1
1
2
3
5
8
13
21
34


'The last one is F9'

斐波那契数列

n-th Fibonacci number $F_n = F_{n-1} + F_{n-2}$ 初始值 $F_0 = 0, F_1 = 1$

In [23]:
def fib(n):
    i, a, b = 0, 0, 1
    while i < n:
        yield b
        a, b = b, a + b
        i += 1
    return 'The last one is F' + str(n)

fib_gen = fib(9)
print(fib_gen)

print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))

<generator object fib at 0x7fb82b7479e8>
1
1
2
3
5
8
13
21
34


StopIteration: ignored


每次调用next()，就计算出下一个元素的值

return 有别于 yield: return 完全终止一个函数，但yield 暂停函数并保存现有状态，然后在后续调用中从那里继续执行。

In [24]:
fib_seq = [1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144]

fib_shifted = [2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233]

golden_ratio = [round(a / b, 3) for a, b in zip(fib_shifted, fib_seq)]
print(golden_ratio)

ratio1 = [round(a / b, 3) for a, b in zip(fib_seq, fib_shifted)]
print(ratio1)

ratio2 = [round(a / b, 3) for a, b in zip(fib_seq[:-1], fib_shifted[1:])]
print(ratio2)

ratio3 = [round(a / b, 3) for a, b in zip(fib_seq[:-2], fib_shifted[2:])]
print(ratio3)

import math
ratio0 = [round(math.sqrt(r), 3) for r in ratio1]
print(ratio0)
fib_ratios = [ratio3[-1], ratio2[-1], ratio1[-1], ratio0[-1]]
print(fib_ratios)

[2.0, 1.5, 1.667, 1.6, 1.625, 1.615, 1.619, 1.618, 1.618, 1.618, 1.618]
[0.5, 0.667, 0.6, 0.625, 0.615, 0.619, 0.618, 0.618, 0.618, 0.618, 0.618]
[0.333, 0.4, 0.375, 0.385, 0.381, 0.382, 0.382, 0.382, 0.382, 0.382]
[0.2, 0.25, 0.231, 0.238, 0.235, 0.236, 0.236, 0.236, 0.236]
[0.707, 0.817, 0.775, 0.791, 0.784, 0.787, 0.786, 0.786, 0.786, 0.786, 0.786]
[0.236, 0.382, 0.618, 0.786]


神奇的斐波那契数列如何应用于技术分析

黄金比率 $\phi = \frac{1 + \sqrt{5}}{2} = 1.618...$

黄金分割 $\Phi = \frac{1}{\phi} = \phi - 1 = 0.618... $

黄金比率可以用斐波那契序列相邻元素的商来逼近，$\lim_{n} {F_{n+1}}/{F_{n}} = \phi$

同样地，可以得到三个重要比例 

$\lim_{n} {F_{n}}/{F_{n+1}} = 0.618$

$\lim_{n} {F_{n}}/{F_{n+2}} = 0.382$

$\lim_{n} {F_{n}}/{F_{n+3}} = 0.236$

有时还加上黄金分割的平方根 $\sqrt{\Phi} = 0.786$

生成斐波那契比例 $[23.6\%, 38.2\%, 61.8\%, 78.6\%]$

用于股票技术分析可找到重要支撑或阻力线。比如当价格大幅上涨时，通常会趋向于按斐波那契比例回撤，然后继续沿原方向上涨。

背后原因属于市场心理/行为学范畴。

In [25]:
fib_seq = [1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144]
PHI = 1.618
print([int(round(fn * PHI)) for fn in fib_seq])
fib_gen = (int(round(fn * PHI)) for fn in fib_seq)
print(fib_gen)

for fn in fib_gen:
  print(fn)

[2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233]
<generator object <genexpr> at 0x7fb82b747780>
2
3
5
8
13
21
34
55
89
144
233


用() 代替 [ ] 就可以简单地把list 变成 generator expression 生成器表达式，即匿名生成器
